### PR TITLE
JS: Env Injection query

### PR DESCRIPTION
**Pull request by am0o0**
Detect user-controllable environment variable injection that can lead to security issues.

### CODE REVIEW

**Comment by ghsecuritylab** [comment]
Hello am0o0 👋
You have submitted this pull request as a bug bounty report in the [github/securitylab](https://github.com/github/securitylab) repository and therefore this pull request has been put into draft state to give time for the GitHub Security Lab to assess the PR. When GitHub Security Lab has finished assessing your pull request, it will be marked automatically as `Ready for review`. Until then, please don't change the draft state.

In the meantime, feel free to make changes to the pull request. If you'd like to maximize payout for your this and future submissions, here are a few general guidelines, that we might take into consideration when reviewing a submission.

- the submission models widely-used frameworks/libraries
- the vulnerability modeled in the submission is impactful
- the submission finds new true positive vulnerabilities
- the submission finds very few false positives
- code in the submission is easy to read and will be easy to maintain
- documentation is written clearly, highlighting the impact of the issue it finds and is written without grammatical or other errors. The code samples clearly show the vulnerability
- the submission includes tests, change note etc.

Please note that these are guidelines, not rules. Since we have a lot of different types of submissions, the guidelines might vary for each submission.

Happy hacking!

File: javascript/ql/src/experimental/Security/CWE-099/EnvValueAndKeyInjection.ql [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  }
+}
+
+from
+  Configuration cfg, Configuration cfg2, DataFlow::PathNode source, DataFlow::PathNode sink1,
+  DataFlow::PathNode sink2
+where
+  cfg.hasFlowPath(source, sink1) and
+  sink1.getNode() = API::moduleImport("process").getMember("env").getAMember().asSink() and
+  cfg2.hasFlowPath(source, sink2) and
```
Comment (erik-krogh) [inline javascript/ql/src/experimental/Security/CWE-099/EnvValueAndKeyInjection.ql:54]: You're restricting your query to only find results where the same source flows to both the value and key of environment variable write.  
If you simple set a `_` here instead, then you'll lift that restriction. 
```suggestion
  cfg2.hasFlowPath(_, sink2) and
```

**Review by erik-krogh — COMMENTED** [review]
You're restricting your query to only find results where the same source flows to both the value and key of environment variable write. 
The sources should not be shared between the two dataflow configurations. 

But you're not restricting that it's the same write to `process.env`.  
So you can have one write like: `process.env[tainted] = "constant"`, and another like: `process.env.constant = tainted`, and the second write would get flagged because the first exist somewhere else.  

You should make a predicate that relate the key and value of the write to `process.env`, and then use that to make sure it's the same write to `process.env` you're finding. 
Something like: 

```CodeQL
private predicate readToProcessEnv(DataFlow::Node key, DataFlow::Node value) {
    exists(DataFlow::PropWrite env | env = NodeJSLib::process().getAPropertyWrite("env") |
      key = env.getPropertyNameExpr().flow() and
      value = env.getRhs()
    )
  }
```

File: javascript/ql/src/experimental/Security/CWE-099/EnvValueAndKeyInjection.qhelp [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+		<p>
+			The following example allows unauthorized users to assign a value to any environment variable.
+		</p>
+
+		<sample src="examples/Bad_Value_And_Key_Assignment" />
+
+	</example>
+
+	<references>
+		<a href="https://huntr.com/bounties/00ec6847-125b-43e9-9658-d3cace1751d6/">Admin account TakeOver in mintplex-labs/anything-llm</a>
```
Comment (erik-krogh) [inline javascript/ql/src/experimental/Security/CWE-099/EnvValueAndKeyInjection.qhelp:33]: You need to wrap this in a `<li>...</li>`, that's why the QHelp renderer is failing. 

Same in `EnvValueInjection.qhelp`.

**Review by erik-krogh — APPROVED** [review]

**Pull request merged**